h2o-core/src/main/java/water/jdbc/SQLManager.java

```diff
@@ -171,49 +183,71 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
+      final int num_retrieval_chunks = ConnectionPoolProvider.estimateConcurrentConnections(H2O.getCloudSize(), H2O.ARGS.nthreads);
+      vec = num_retrieval_chunks >= num_chunks
+              ? Vec.makeCon(numRow, num_chunks)
+              : Vec.makeConN(numRow, num_retrieval_chunks);
     } else {
-      double rows_per_chunk = FileVec.calcOptimalChunkSize(totSize, numCol, numCol * 4,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false, false);
-      _v = makeCon(0, numRow, (int) Math.ceil(Math.log1p(rows_per_chunk)), false);
+      vec = Vec.makeCon(numRow, num_chunks);
     }
-    Log.info("Number of chunks: " + _v.nChunks());
+    //if autoRebalance set to true, we first use optimal #chunks for retrieval and then immediately rebalance to optimal #chunks for later data processing
+    final boolean autoRebalance = vec.nChunks() != num_chunks && Boolean.parseBoolean(System.getProperty(AUTO_REBALANCE_ENABLED, "false"));
+    Log.info("Number of chunks for data retrieval: " + vec.nChunks());
+    Log.info("Number of final chunks: " + (autoRebalance ? vec.nChunks() : vec.nChunks()));
     //create frame
-    final Key destination_key = Key.make(table + "_sql_to_hex");
+    final Key destination_key = Key.make((table + "_sql_to_hex").replaceAll("\\W", "_"));
     final Job<Frame> j = new Job(destination_key, Frame.class.getName(), "Import SQL Table");
 
     final String finalTable = table;
     H2O.H2OCountedCompleter work = new H2O.H2OCountedCompleter() {
       @Override
       public void compute2() {
-        ConnectionPoolProvider provider = new ConnectionPoolProvider(connection_url, username, password, _v.nChunks());
-        Frame fr = new SqlTableToH2OFrame(finalTable, needFetchClause, columns, numCol, j, provider).doAll(columnH2OTypes, _v)
+        final ConnectionPoolProvider provider = new ConnectionPoolProvider(connection_url, username, password, vec.nChunks());
+        final Frame retrieval_fr = new SqlTableToH2OFrame(finalTable, needFetchClause, columns, numCol, j, provider)
+                .doAll(columnH2OTypes, vec)
                 .outputFrame(destination_key, columnNames, null);
+
+        Frame fr = retrieval_fr;
+
+        if (autoRebalance) {
+          final RebalanceDataSet rds = new RebalanceDataSet(retrieval_fr, destination_key, num_chunks);
+          H2O.submitTask(rds);
+          fr = rds.getResult();
+          retrieval_fr.remove();
+        }
+
+        vec.remove();
+
         DKV.put(fr);
-        _v.remove();
         ParseDataset.logParseResults(fr);
-        if (finalTable.equals(SQLManager.TEMP_TABLE_NAME)) 
+        if (finalTable.equals(SQLManager.TEMP_TABLE_NAME))
           dropTempTable(connection_url, username, password);
         tryComplete();
       }
     };
-    j.start(work, _v.nChunks());
+    j.start(work, vec.nChunks());
     
     return j;
   }
 
   static class ConnectionPoolProvider extends Iced<ConnectionPoolProvider> {
 
-    private String _url;
```

Comment:
I would rather respect h2o coding conventions

Reply:
where can I find those? the codebase itself is very inconsistent.
you mentioned in a previous comment: "I would remove the leading underscore from _v and _retrieval_v. It goes against the coding conventions and this is a good opportunity to fix it."
so, I removed leading underscores everywhere in this file.
Do you mean that it should stay for private fields?
Btw, those renaming are not really related with my changes, I'd rather use the refactoring branch (PR 2667) for that.

Reply:
The conventions are not formalized.

The one rule we follow is to use underscores in the name of class fields. That distinguishes them from method variables. In your case `_v` was a variable, here we are dealing with class fields which already begin with underscore as expected.

Reply:
changed